gvisor: init at 2019-11-14

andrew-d opened the pull request:
###### Motivation for this change
This is a revamp of #50218 after enough upstream changes that it's possible to build inside a Nix sandbox. This was requested in #39889, but there were some problems with Bazel at the time. I've managed to get this working with `buildBazelPackage`. At the end of the whole process, gvisor is runnable:

```
$ /nix/store/7k1c1jikms1pjimk8561x18xpj51dm5l-gvisor-2019-11-08/bin/runsc --help
Usage: runsc <flags> <subcommand> <subcommand args>

Subcommands:
	checkpoint       checkpoint current state of container (experimental)
	create           create a secure container
	delete           delete resources held by a container
	do               Simplistic way to execute a command inside the sandbox. It's to be used for testing only.
	events           display container events such as OOM notifications, cpu, memory, and IO usage statistics
```

(I also added the `containerd` shim as well, since it didn't feel worth another PR)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] ~macOS~
   - [ ] other Linux distributions
- [x] ~Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))~
- [x] ~Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`~
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc original reviewers @nlewo, @Profpatsch, and commenters @dtzWill, @benpye and @ghuntley  

Closes #50218
Closes #39889

andrew-d committed:
gvisor: init at 2019-11-14

andrew-d committed:
gvisor-containerd-shim: init at 2019-10-09

Profpatsch commented:
Do you have sandboxing enabled in your local nix? The build fails, because it tries to download some repository dependencies. Most of the nix world has sandboxing enabled, which restricts network access to fixed-output derivations.

andrew-d commented:
@Profpatsch - I do, yes. I think the problem is that the `buildBazelPackage` tools remove `rules_cc`, which is now an external dependency and then can't be downloaded. There's also some issues with rules_go and [`x/net/nettest`](https://godoc.org/golang.org/x/net/nettest) that I'm trying to debug as well.

andrew-d commented:
@Profpatsch - Okay, between #74013 and this PR, it's building successfully and reproducibly :tada:

Profpatsch commented on pkgs/build-support/build-bazel-package/default.nix:
```diff
@@ -4,10 +4,20 @@
+, bazelTarget
+, buildAttrs
+, fetchAttrs
+, removeRulesCC ? true
```
Can you document this argument, please? it seems to be a bit specific to the gvisor use-case?

andrew-d replied:
Done in #74013

flokli commented:
Can you add a simple nixos vm test starting a container, so we can verify it works?

andrew-d committed:
nixosTests.gvisor: add test for gvisor and gvisor as a Docker runtime

andrew-d commented:
@flokli - Okay, added a test that exercises both the `gvisor do` subcommand and gvisor when it's being used as a Docker runtime. I hadn't previously tested the `gvisor do` subcommand, so I also added a patch to ensure that the path to `ip`/`ipconfig`/`sysctl` is absolute.

andrew-d commented:
@Profpatsch / @flokli - Okay, rebased now that #74103 is merged. This should be ready to merge now, and tests pass locally.

flokli commented on pkgs/applications/virtualization/gvisor/default.nix:
```diff
@@ -0,0 +1,107 @@
+      WORKSPACE
+
+    # The gazelle Go tooling needs CA certs
+    export SSL_CERT_FILE="${cacert}/etc/ssl/certs/ca-bundle.crt"
```
This is used during the fetcher phase of `buildBazelPackage`. I assume it's not baked into the binary, right?

Quickly looking through nixpkgs, it seems `pkgs/build-support/build-bazel-package/default.nix` already sets `GIT_SSL_CAINFO=`. Isn't that not enough? What do you think about setting `SSL_CERT_FILE` in there, too, instead of here?

andrew-d replied:
I could do that; opened as #74536 if that's something you'd want to merge.

The reason that `GIT_SSL_CAINFO` isn't enough is because it's not just `git` fetching things; `rules_go` runs a Go binary that fetches sources, which means that it needs to know about SSL certs.  If you're curious: the Go tooling makes a request with `?go-get=1` to see where it should fetch a source from; more details [here](https://golang.org/cmd/go/).

flokli replied:
Thanks for the explanation :-)

#74536 was merged - can you remove the line from here?

andrew-d replied:
@flokli - Thanks for the quick merge! Removed this line and rebased on `master`.

flokli commented on pkgs/applications/virtualization/gvisor/default.nix:
```diff
@@ -0,0 +1,107 @@
+, buildBazelPackage
+, fetchFromGitHub
+, cacert
+
```
can you remove that stray newline?

andrew-d replied:
Done. I will need to rebase again if #74536 is merged, but did this now.

flokli commented on pkgs/applications/virtualization/gvisor/default.nix:
```diff
@@ -0,0 +1,103 @@
+    # available in the $PATH; patch these now.
+    patches = [ ./gvisor-do-paths.patch ];
+    postPatch = ''
+      substituteInPlace runsc/cmd/do.go \
```
For the sake of maintainability:

Could we just do a `wrapProgram $out/bin/runsc --prefix PATH : ${stdenv.lib.makeBinPath [ iproute iptables procps ]}` in `installPhase`, instead of patching the source code?

andrew-d replied:
@flokli - Done! And confirmed that it works in the NixOS test as well.

andrew-d commented:
I also just bumped to an actual tagged release of `gvisor`, since they tagged [`release-20191114.0`](https://github.com/google/gvisor/releases/tag/release-20191114.0) after I'd opened this PR.

flokli commented:
Thanks!
